File: src/Compilers/VisualBasic/Test/Emit/PDB/TypeDefinitionDocumentTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.IO
Imports System.Reflection.Metadata
Imports System.Reflection.Metadata.Ecma335
Imports System.Text
Imports Microsoft.CodeAnalysis.Debugging
Imports Microsoft.CodeAnalysis.Emit
Imports Microsoft.CodeAnalysis.Test.Utilities
Imports Roslyn.Test.Utilities

Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests.PDB
    Public Class TypeDefinitionDocumentTests
        Inherits BasicTestBase

        <Fact>
        Public Sub ClassWithMethod()
            Dim source As String = "
Class M
    Public Shared Sub A()
        System.Console.WriteLine()
    End Sub
End Class
"
            TestTypeDefinitionDocuments({source})
        End Sub

        <Fact>
        Public Sub NestedClassWithMethod()
            Dim source As String = "
Class C
    Class N
        Public Shared Sub A()
            System.Console.WriteLine()
        End Sub
    End Class
End Class
"
            TestTypeDefinitionDocuments({source})
        End Sub

        <Fact>
        Public Sub MultiNestedClassWithMethod()
            Dim source As String = "
Class C
    Class N
        Class N2
            Public Shared Sub A()
                System.Console.WriteLine()
            End Sub
        End Class
    End Class
End Class
"
            TestTypeDefinitionDocuments({source})
        End Sub

        <Fact>
        Public Sub PartialMultiNestedClassWithMethod()
            Dim source1 As String = "
Partial Class C
    Partial Class N
        Public Shared Sub A()
            System.Console.WriteLine()
        End Sub
    End Class
End Class
"

            Dim source2 As String = "
Partial Class C
    Partial Class N
    End Class
End Class
"
            TestTypeDefinitionDocuments({source1, source2},
                              ("C", "2.vb"))
        End Sub

        <Fact>
        Public Sub EmptyClass()
            Dim source As String = "
Class O
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("O", "1.vb"))
        End Sub

        <Fact>
        Public Sub EmptyNestedClass()
            Dim source As String = "
Class O
    Class N
    End Class
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("O", "1.vb"))
        End Sub

        <Fact>
        Public Sub EmptyMultiNestedClass()
            Dim source As String = "
Class O
    Class N
        Class N2
        End Class
    End Class
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("O", "1.vb"))
        End Sub

        <Fact>
        Public Sub MultipleClassesAndFiles()
            Dim source1 As String = "
Class M
    Public Shared Sub A()
        System.Console.WriteLine()
    End Sub
End Class

Class N
End Class

Class O
End Class
"
            Dim source2 As String = "
Class C
End Class

Class D
End Class
"
            TestTypeDefinitionDocuments({source1, source2},
                              ("N", "1.vb"),
                              ("O", "1.vb"),
                              ("C", "2.vb"),
                              ("D", "2.vb"))
        End Sub

        <Fact>
        Public Sub PartialClasses()
            Dim source1 As String = "
Partial Class C
End Class
"
            Dim source2 As String = "
Partial Class C
End Class
"
            TestTypeDefinitionDocuments({source1, source2},
                              ("C", "1.vb, 2.vb"))
        End Sub

        <Fact>
        Public Sub PartialClasses2()
            Dim source1 As String = "
Partial Class C
End Class
"
            Dim source2 As String = "
Partial Class C
    Private x As Integer

    Sub M()
    End Sub
End Class
"
            TestTypeDefinitionDocuments({source1, source2},
                              ("C", "1.vb"))
        End Sub

        <Fact>
        Public Sub PartialClasses3()
            Dim source1 As String = "
Partial Class C
End Class
"
            Dim source2 As String = "
Partial Class C
    Private x As Integer = 1

    Sub M()
    End Sub
End Class
"
            TestTypeDefinitionDocuments({source1, source2},
                              ("C", "1.vb"))
        End Sub

        <Fact>
        Public Sub [Property]()
            Dim source As String = "
Class C
    Public Property X As Integer
End Class

Class D
    Public Property Y As Integer = 4
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("C", "1.vb"))
        End Sub

        <Fact>
        Public Sub Fields()
            Dim source As String = "
Class C
    Private x As Integer
    Private Const y As Integer = 3
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("C", "1.vb"))
        End Sub

        <Fact>
        Public Sub Fields_WithInitializer()
            Dim source As String = "
Class C
    Private x As Integer = 1
    Private Const y As Integer = 3
End Class
"
            TestTypeDefinitionDocuments({source})
        End Sub

        <Fact>
        Public Sub AbstractMethod()
            Dim source As String = "
MustInherit Class C
    MustOverride Sub M()
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("C", "1.vb"))
        End Sub

        <Fact>
        Public Sub Interfaces()
            Dim source1 As String = "
Interface I1
End Interface

Partial Interface I2
    Sub M()
End Interface
"
            Dim source2 As String = "
Partial Interface I2
End Interface
"
            TestTypeDefinitionDocuments({source1, source2},
                              ("I1", "1.vb"),
                              ("I2", "1.vb, 2.vb"))
        End Sub

        <Fact>
        Public Sub [Enum]()
            Dim source As String = "
Enum E
    A
    B
End Enum
"
            TestTypeDefinitionDocuments({source},
                              ("E", "1.vb"))
        End Sub

        <Fact>
        Public Sub [Delegate]()
            Dim source As String = "
Delegate Sub D(a As Integer)

Class C
    Public Sub M()
        Dim y = (
        Function(ByRef a As Integer)
            Return a
        End Function
        )
    End Sub
End Class
"
            TestTypeDefinitionDocuments({source},
                              ("D", "1.vb"))
        End Sub

        <Fact>
        Public Sub AnonymousType()
            Dim source As String = "
Public Class C
    Public Sub M()
       Dim x = New With { .Goo = 1, .Bar = ""Hi"" }
    End Sub
End Class
"
            TestTypeDefinitionDocuments({source})
        End Sub

        <Fact>
        Public Sub ExternalSourceDirectives()
            Dim source As String = "
Class C
#ExternalSource (""C.vb"", 1)
    Public Sub M()
    End Sub
#End ExternalSource
End Class

Class D
#ExternalSource (""D.vb"", 1)
    Private _x As Integer = 1
    Private Property Y As Integer = 1
#End ExternalSource
End Class

#ExternalSource (""E.vb"", 1)
Class E
    Public Sub M()
    End Sub
End Class
#End ExternalSource

#ExternalSource (""F.vb"", 1)
Class F
End Class
#End ExternalSource
"
            TestTypeDefinitionDocuments({source},
                              ("C", "1.vb"),
                              ("D", "1.vb"),
                              ("E", "1.vb"),
                              ("F", "1.vb"))
        End Sub

        Public Shared Sub TestTypeDefinitionDocuments(sources As String(), ParamArray expected As (String, String)())
            Dim trees = sources.Select(Function(s, i) SyntaxFactory.ParseSyntaxTree(s, path:=$"{i + 1}.vb", encoding:=Encoding.UTF8)).ToArray()
            Dim compilation = CreateCompilation(trees, options:=TestOptions.DebugDll)

            Dim pdbStream = New MemoryStream()
            Dim pe = compilation.EmitToArray(EmitOptions.[Default].WithDebugInformationFormat(DebugInformationFormat.PortablePdb), pdbStream:=pdbStream)
            pdbStream.Position = 0

            Dim metadata = ModuleMetadata.CreateFromImage(pe)
            Dim metadataReader = metadata.GetMetadataReader()

            Using provider = MetadataReaderProvider.FromPortablePdbStream(pdbStream)
                Dim pdbReader = provider.GetMetadataReader()

                Dim actual = From handle In pdbReader.CustomDebugInformation
                             Let entry = pdbReader.GetCustomDebugInformation(handle)
                             Where pdbReader.GetGuid(entry.Kind).Equals(PortableCustomDebugInfoKinds.TypeDefinitionDocuments)
                             Select (GetTypeName(metadataReader, entry.Parent), GetDocumentNames(pdbReader, entry.Value))

                AssertEx.Equal(expected, actual, itemSeparator:=", ", itemInspector:=Function(i) $"(""{i.Item1}"", ""{i.Item2}"")")
            End Using
        End Sub

        Private Shared Function GetTypeName(metadataReader As MetadataReader, handle As EntityHandle) As String
            Dim typeHandle = CType(handle, TypeDefinitionHandle)
            Dim type = metadataReader.GetTypeDefinition(typeHandle)
            Return metadataReader.GetString(type.Name)
        End Function

        Private Shared Function GetDocumentNames(pdbReader As MetadataReader, value As BlobHandle) As String
            Dim result = New List(Of String)()
            Dim reader = pdbReader.GetBlobReader(value)

            While reader.RemainingBytes > 0
                Dim documentRow = reader.ReadCompressedInteger()

                If documentRow > 0 Then
                    Dim doc = pdbReader.GetDocument(MetadataTokens.DocumentHandle(documentRow))
                    result.Add(pdbReader.GetString(doc.Name))
                End If
            End While

            ' Order can be different in net472 vs net5 :(
            result.Sort()
            Return String.Join(", ", result)
        End Function
    End Class
End Namespace

File: src/Compilers/CSharp/Test/Emit/PDB/TypeDefinitionDocumentTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection.Metadata;
using System.Reflection.Metadata.Ecma335;
using System.Text;
using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
using Microsoft.CodeAnalysis.Debugging;
using Microsoft.CodeAnalysis.Emit;
using Microsoft.CodeAnalysis.Test.Utilities;
using Roslyn.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.CSharp.UnitTests.PDB
{
    public class TypeDefinitionDocumentTests : CSharpTestBase
    {
        [Fact]
        public void ClassWithMethod()
        {
            string source = @"
class M
{
    public static void A()
    {
        System.Console.WriteLine();
    }
}
";
            TestTypeDefinitionDocuments(new[] { source });
        }

        [Fact]
        public void NestedClassWithMethod()
        {
            string source = @"
class C
{
    class N
    {
        public static void A()
        {
            System.Console.WriteLine();
        }
    }
}
";
            TestTypeDefinitionDocuments(new[] { source });
        }

        [Fact]
        public void MultiNestedClassWithMethod()
        {
            string source = @"
class C
{
    class N
    {
        class N2
        {
            public static void A()
            {
                System.Console.WriteLine();
            }
        }
    }
}
";
            TestTypeDefinitionDocuments(new[] { source });
        }

        [Fact]
        public void PartialNestedClassWithMethod()
        {
            string source1 = @"
partial class C
{
    partial class N
    {
        public static void A()
        {
            System.Console.WriteLine();
        }
    }
}
";
            string source2 = @"
partial class C
{
    partial class N
    {
    }
}
";
            TestTypeDefinitionDocuments(new[] { source1, source2 },
                ("C", "2.cs"));
        }

        [Fact]
        public void EmptyClass()
        {
            string source = @"
class O
{
}
";
            TestTypeDefinitionDocuments(new[] { source },
                ("O", "1.cs"));
        }

        [Fact]
        public void EmptyNestedClass()
        {
            string source = @"
class O
{
    class N
    {
    }
}
";
            TestTypeDefinitionDocuments(new[] { source },
                ("O", "1.cs"));
        }

        [Fact]
        public void EmptyMultiNestedClass()
        {
            string source = @"
class O
{
    class N
    {
        class N2
        {
        }
    }
}
";
            TestTypeDefinitionDocuments(new[] { source },
                ("O", "1.cs"));
        }

        [Fact]
        public void MultipleClassesAndFiles()
        {
            string source1 = @"
class M
{
    public static void A()
    {
        System.Console.WriteLine();
    }
}

class N
{
}

class O
{
}
";

            string source2 = @"
class C
{
}

class D
{
}
";

            TestTypeDefinitionDocuments(new[] { source1, source2 },
                ("N", "1.cs"),
                ("O", "1.cs"),
                ("C", "2.cs"),
                ("D", "2.cs"));
        }

        [Fact]
        public void PartialClasses()
        {
            string source1 = @"
partial class C
{
}
";
            string source2 = @"
partial class C
{
}
";
            TestTypeDefinitionDocuments(new[] { source1, source2 },
                ("C", "1.cs, 2.cs"));
        }

        [Fact]
        public void PartialClasses2()
        {
            string source1 = @"
partial class C
{
}
";
            string source2 = @"
partial class C
{
    int x = 1;
    void M() { }
}
";
            TestTypeDefinitionDocuments(new[] { source1, source2 },
                ("C", "1.cs"));
        }

        [Fact]
        public void PartialClasses3()
        {
            string source1 = @"
partial class C
{
}
";
            string source2 = @"
partial class C
{
    int x;
    void M() { }
}
";
            TestTypeDefinitionDocuments(new[] { source1, source2 },
                ("C", "1.cs"));
        }

        [Fact]
        public void Property()
        {
            string source = @"
class C
{
    public int X { get; set; }
}
";

            TestTypeDefinitionDocuments(new[] { source });
        }

        [Fact]
        public void Fields()
        {
            string source = @"
class C
{
    int x;
    const int z = 2;
    int y;
}
";

            TestTypeDefinitionDocuments(new[] { source },
                ("C", "1.cs"));
        }

        [Fact]
        public void Fields_WithInitializer()
        {
            string source = @"
class C
{
    int x;
    const int z = 2;
    int y = 1;
}
";

            TestTypeDefinitionDocuments(new[] { source });
        }

        [Fact]
        public void AbstractMethod()
        {
            string source = @"
abstract class C
{
    public abstract void M();
}
";

            TestTypeDefinitionDocuments(new[] { source },
                ("C", "1.cs"));
        }

        [Fact]
        public void ExternMethod()
        {
            string source = @"
class C
{
    public extern void M();
}
";

            TestTypeDefinitionDocuments(new[] { source },
                ("C", "1.cs"));
        }

        [Fact]
        public void Interfaces()
        {
            string source1 = @"
interface I1
{
}

partial interface I2
{
    public void F();
}
";
            string source2 = @"
partial interface I2
{
}
";
            TestTypeDefinitionDocuments(new[] { source1, source2 },
                ("I1", "1.cs"),
                ("I2", "1.cs, 2.cs"));
        }

        [Fact]
        public void Record()
        {
            string source = @"
record R(int X);
";

            // The compiler synthesized methods have document info so we don't expect a type document
            TestTypeDefinitionDocuments(new[] { source, IsExternalInitTypeDefinition },
                ("IsExternalInit", "2.cs"));
        }

        [Fact]
        public void Record_SynthesizedMember()
        {
            string source = @"
record R(int X)
{
    protected virtual bool PrintMembers(System.Text.StringBuilder builder)
    {
        return true;
    }
}
";

            TestTypeDefinitionDocuments(new[] { source, IsExternalInitTypeDefinition },
                ("IsExternalInit", "2.cs"));
        }

        [Fact]
        public void Enum()
        {
            string source = @"
enum E
{
}

enum E2
{
    A,
    B
}
";

            TestTypeDefinitionDocuments(new[] { source },
                ("E", "1.cs"),
                ("E2", "1.cs"));
        }

        [Fact]
        public void Delegate()
        {
            string source = @"
delegate void D(int a);

class C
{
    void M()
    {
        var x = (int a, ref int b) => a;
    }
}
";

            TestTypeDefinitionDocuments(new[] { source },
                ("D", "1.cs"));
        }

        [Fact]
        public void AnonymousTypes()
        {
            string source = @"
class C
{
    void M()
    {
        var x = new { Goo = 1, Bar = ""Hi"" };
    }
}
";

            TestTypeDefinitionDocuments(new[] { source });
        }

        [Fact]
        public void LineDirectives()
        {
            string source = @"
class C
{
#line 1 ""C.cs""
    void M()
    {
    }
#line default
}

class D
{
#line 1 ""D.cs""
    private int _x = 1;
    private int X { get; set; } = 1;
}

#line 1 ""E.cs""
class E
{
    void M()
    {
    }
}

#line 1 ""F.cs""
class F
{
}
";

            TestTypeDefinitionDocuments(new[] { source },
                ("C", "1.cs"),
                ("D", "1.cs"),
                ("E", "1.cs"),
                ("F", "1.cs"));
        }

        private static void TestTypeDefinitionDocuments(string[] sources, params (string typeName, string documentName)[] expected)
        {
            var trees = sources.Select((s, i) => SyntaxFactory.ParseSyntaxTree(s, path: $"{i + 1}.cs", encoding: Encoding.UTF8)).ToArray();
            var compilation = CreateCompilation(trees, options: TestOptions.DebugDll);

            var pdbStream = new MemoryStream();
            var pe = compilation.EmitToArray(EmitOptions.Default.WithDebugInformationFormat(DebugInformationFormat.PortablePdb), pdbStream: pdbStream);
            pdbStream.Position = 0;

            var metadata = ModuleMetadata.CreateFromImage(pe);
            var metadataReader = metadata.GetMetadataReader();

            using var provider = MetadataReaderProvider.FromPortablePdbStream(pdbStream);
            var pdbReader = provider.GetMetadataReader();

            var actual = from handle in pdbReader.CustomDebugInformation
                         let entry = pdbReader.GetCustomDebugInformation(handle)
                         where pdbReader.GetGuid(entry.Kind).Equals(PortableCustomDebugInfoKinds.TypeDefinitionDocuments)
                         select (typeName: GetTypeName(entry.Parent), documentName: GetDocumentNames(entry.Value));

            AssertEx.Equal(expected, actual, itemSeparator: ",\n", itemInspector: i => $"(\"{i.typeName}\", \"{i.documentName}\")");

            string GetTypeName(EntityHandle handle)
            {
                var typeHandle = (TypeDefinitionHandle)handle;
                var type = metadataReader.GetTypeDefinition(typeHandle);
                return metadataReader.GetString(type.Name);
            }

            string GetDocumentNames(BlobHandle value)
            {
                var result = new List<string>();

                var reader = pdbReader.GetBlobReader(value);
                while (reader.RemainingBytes > 0)
                {
                    var documentRow = reader.ReadCompressedInteger();
                    if (documentRow > 0)
                    {
                        var doc = pdbReader.GetDocument(MetadataTokens.DocumentHandle(documentRow));
                        result.Add(pdbReader.GetString(doc.Name));
                    }
                }

                return string.Join(", ", result);
            }
        }
    }
}
